### PR TITLE
feat: add versioning to agent config

### PR DESCRIPTION
**Pull request by fontanierh**
- adds a version field to `AgentConfiguration`
- makes `AgentConfiguration` unique on `(sId, version)`
- adds a migration to remove unicity on `(sId)`
- adds a `agentConfigurationVersion` field to `AgentMessage`
- update the `getAgent` logic to fetch the agent that has the highest version
- add the ability to pass the `sId` (default to unique new sId) to the agent creation code. If the sId is passed, the the new agent will have an increamented version for that sId and existing assistants with that sId will be archived

The unique indexes on `sId` from `AgentConfiguration` were duplicated multiple times (for some reason), which is why the SQL migration is dropping multiple constraints. see [here](https://metabase.dust.tt/question#eyJkYXRhc2V0X3F1ZXJ5Ijp7InR5cGUiOiJuYXRpdmUiLCJuYXRpdmUiOnsicXVlcnkiOiJTRUxFQ1QgY29uLipcbiAgICAgICBGUk9NIHBnX2NhdGFsb2cucGdfY29uc3RyYWludCBjb25cbiAgICAgICAgICAgIElOTkVSIEpPSU4gcGdfY2F0YWxvZy5wZ19jbGFzcyByZWxcbiAgICAgICAgICAgICAgICAgICAgICAgT04gcmVsLm9pZCA9IGNvbi5jb25yZWxpZFxuICAgICAgICAgICAgSU5ORVIgSk9JTiBwZ19jYXRhbG9nLnBnX25hbWVzcGFjZSBuc3BcbiAgICAgICAgICAgICAgICAgICAgICAgT04gbnNwLm9pZCA9IGNvbm5hbWVzcGFjZVxuICAgICAgIFdIRVJFIHJlbC5yZWxuYW1lID0gJ2FnZW50X2NvbmZpZ3VyYXRpb25zJzsiLCJ0ZW1wbGF0ZS10YWdzIjp7fX0sImRhdGFiYXNlIjo0fSwiZGlzcGxheSI6InRhYmxlIiwicGFyYW1ldGVycyI6W10sInZpc3VhbGl6YXRpb25fc2V0dGluZ3MiOnt9fQ==)

PATCH code has  been updated to use versioning as well, and the "update" methods were deleted. "create" methods for action / generation / agent configurations have been taken out of the interface, in favour of a `createOrUpgradeAgentConfiguration` method that does all the creates and handles the versioning

`getAgentConfigurations` has also been fixed  to not look at assistants that are archived.

⚠️ In order:
- run SQL migration
- run initdb
- deploy

### CODE REVIEW

File: front/migrations/20230920_agent_config_drop_sid_unique_constraints.sql [hunk 1]
```diff
@@ -0,0 +1,17 @@
+alter table
```
Comment (fontanierh) [inline front/migrations/20230920_agent_config_drop_sid_unique_constraints.sql:1]: see [here](https://metabase.dust.tt/question#eyJkYXRhc2V0X3F1ZXJ5Ijp7InR5cGUiOiJuYXRpdmUiLCJuYXRpdmUiOnsicXVlcnkiOiJTRUxFQ1QgY29uLipcbiAgICAgICBGUk9NIHBnX2NhdGFsb2cucGdfY29uc3RyYWludCBjb25cbiAgICAgICAgICAgIElOTkVSIEpPSU4gcGdfY2F0YWxvZy5wZ19jbGFzcyByZWxcbiAgICAgICAgICAgICAgICAgICAgICAgT04gcmVsLm9pZCA9IGNvbi5jb25yZWxpZFxuICAgICAgICAgICAgSU5ORVIgSk9JTiBwZ19jYXRhbG9nLnBnX25hbWVzcGFjZSBuc3BcbiAgICAgICAgICAgICAgICAgICAgICAgT04gbnNwLm9pZCA9IGNvbm5hbWVzcGFjZVxuICAgICAgIFdIRVJFIHJlbC5yZWxuYW1lID0gJ2FnZW50X2NvbmZpZ3VyYXRpb25zJzsiLCJ0ZW1wbGF0ZS10YWdzIjp7fX0sImRhdGFiYXNlIjo0fSwiZGlzcGxheSI6InRhYmxlIiwicGFyYW1ldGVycyI6W10sInZpc3VhbGl6YXRpb25fc2V0dGluZ3MiOnt9fQ==)

File: front/lib/api/assistant/configuration.ts [hunk 2]
```diff
@@ -47,10 +47,27 @@ export async function getAgentConfiguration(
     return await getGlobalAgent(auth, agentId);
   }
 
+  const latestVersion = await AgentConfiguration.max<
```
Comment (spolu) [inline front/lib/api/assistant/configuration.ts:50]: We can surely do it in one query no?

File: front/lib/api/assistant/configuration.ts [hunk 3]
```diff
@@ -200,23 +218,55 @@ export async function createAgentConfiguration(
...
   }
 ): Promise<AgentConfigurationType> {
   const owner = auth.workspace();
   if (!owner) {
     throw new Error("Unexpected `auth` without `workspace`.");
   }
 
+  let version = 0;
+
+  if (sId) {
```
Comment (spolu) [inline front/lib/api/assistant/configuration.ts:240]: This should be protected by a transaction?

File: front/lib/api/assistant/conversation.ts [hunk 4]
```diff
@@ -767,6 +768,7 @@ export async function* retryAgentMessage(
       {
         status: "created",
         agentConfigurationId: messageRow.agentMessage.agentConfigurationId,
+        agentConfigurationVersion: 0,
```
Comment (spolu) [inline front/lib/api/assistant/conversation.ts:771]: ?

**Review by spolu — CHANGES_REQUESTED** [review]

File: front/migrations/20230920_agent_config_drop_sid_unique_constraints.sql (hunk 1 above)
Comment (PopDaph) [reply front/migrations/20230920_agent_config_drop_sid_unique_constraints.sql:1]: Thank you 🙏🏻

File: front/lib/api/assistant/configuration.ts (hunk 2 above)
Comment (fontanierh) [reply front/lib/api/assistant/configuration.ts:50]: With sequelize I couldn't find how, will look again

File: front/lib/api/assistant/conversation.ts (hunk 4 above)
Comment (fontanierh) [reply front/lib/api/assistant/conversation.ts]: my bad this was for testing forgot to remove

File: front/lib/api/assistant/configuration.ts (hunk 2 above)
Comment (fontanierh) [reply front/lib/api/assistant/configuration.ts:50]: found a solution

Comment (fontanierh) [reply front/lib/api/assistant/configuration.ts]: fixed

**Review by spolu — APPROVED** [review]

File: front/lib/api/assistant/configuration.ts [hunk 5]
```diff
@@ -200,37 +204,81 @@ export async function createAgentConfiguration(
...
+          { status: "archived" },
+          {
+            where: {
+              sId: sId,
+              workspaceId: owner.id,
+            },
+            transaction: t,
+          }
+        );
+      }
```
Comment (PopDaph) [inline front/lib/api/assistant/configuration.ts:254]: NIT: Could be like this to match the query in getAgentConfiguration
```js
const agent = await AgentConfiguration.findOne({
    where: {
      sId: agentId,
      workspaceId: owner.id,
    },
    order: [["version", "DESC"]],
    ],
    limit: 1,
  } , transaction: t );
await agent.update({ status: "archived" }, transaction: t}
```

or would that work? (I'm not sure)
```js
await AgentConfiguration.update(
    { status: 'archived' },
    {
      where: {
        sId: sId,
        workspaceId: owner.id,
        version: {
          [Op.eq]: Sequelize.literal(`(SELECT MAX(version) FROM AgentConfigurations WHERE sId = ${sId} AND workspaceId = ${owner.id})`)
        }
      },
      transaction: t
    }
  );

```

**Review by PopDaph — DISMISSED** [review]
🚀

File: front/lib/api/assistant/configuration.ts (hunk 5 above)
Comment (fontanierh) [reply front/lib/api/assistant/configuration.ts:254]:
```
const agent = await AgentConfiguration.findOne({
    where: {
      sId: agentId,
      workspaceId: owner.id,
    },
    order: [["version", "DESC"]],
    ],
    limit: 1,
  } , transaction: t );
await agent.update({ status: "archived" }, transaction: t}
```

This seems like more work no ? We'd need to fetch the whole config for nothing


For the second one.... I'm not a fan of the raw sql without type checking + I still need the actual version number so I can +1 it to create the new one

File: front/lib/api/assistant/configuration.ts [hunk 6]
```diff
@@ -200,37 +204,81 @@ export async function createAgentConfiguration(
...
     action,
+    sId,
   }: {
     name: string;
     description: string;
     pictureUrl: string;
     status: AgentConfigurationStatus;
     generation: AgentGenerationConfigurationType | null;
     action: AgentActionConfigurationType | null;
+    version?: number;
```
Comment (spolu) [inline front/lib/api/assistant/configuration.ts:215]: version is added here but not above should be removed?

File: front/lib/api/assistant/configuration.ts [hunk 7]
```diff
@@ -200,37 +204,81 @@ export async function createAgentConfiguration(
...
+    sId?: string;
```
Comment (spolu) [inline front/lib/api/assistant/configuration.ts:216]: agentCongiruationId

**Review by spolu — CHANGES_REQUESTED** [review]

File: front/lib/api/assistant/configuration.ts (hunk 5 above)
Comment (PopDaph) [reply front/lib/api/assistant/configuration.ts:254]: Sure! Works as is!

File: front/lib/api/assistant/configuration.ts (hunk 6 above)
Comment (fontanierh) [reply front/lib/api/assistant/configuration.ts:215]: Indeed, a leftover from prev version

File: front/lib/api/assistant/configuration.ts (hunk 7 above)
Comment (fontanierh) [reply front/lib/api/assistant/configuration.ts:216]: Updating 👍

**Comment by spolu** [comment]
Can we remove updateAgentConfiguration entirely in this PR?

**Comment by fontanierh** [comment]
> Can we remove updateAgentConfiguration entirely in this PR?

You mean update the PATCH logic in this PR as well ? Sure can do that if preferred

**Comment by spolu** [comment]
Yes and remove the useless function entirely :+1:

**Review by spolu — CHANGES_REQUESTED** [review]
Let's not move schemas to lib/api

Keep it in POST and import from PATCH (as we've been doing so far).

If you really really really want to dry let's have a function in POST that we use from PATCH but I'm not convinced we need drying here.

**Review by spolu — APPROVED** [review]

**Pull request merged**